File: backend/src/main/java/com/cap/MyLearning/service/UserService.java
package com.cap.MyLearning.service;

import com.cap.MyLearning.model.CourseDetails;
import com.cap.MyLearning.model.User;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UsernameNotFoundException;


import java.util.List;
import java.util.Optional;

public interface UserService {
    List<User> getAllUsers();

    Optional<User> getUserById(Long id);

    User createUser(User user);

    User updateUser(Long id, User user);

    void deleteUser(Long id);

    UserDetails loadUserByUsername(String email) throws UsernameNotFoundException;

    User findByEmail(String email);

    User registerForCourse(Long userId, Long courseId);

    List<CourseDetails> getRegisteredCourseDetails(Long studentId, Long courseId);
}

File: frontend/src/app/services/user.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { User } from '../models/user.model';
import { CourseDetails } from '../models/course-details.model';

@Injectable({
  providedIn: 'root'
})
export class UserService {
  registerToCourseForDate(userId: any, courseId: number, selectedDateTimeId: number) {
    throw new Error('Method not implemented.');
  }

  private apiUrl = 'http://localhost:8080/api/users';

  constructor(private http: HttpClient) { }

  getUsers(): Observable<User[]> {
    return this.http.get<User[]>(this.apiUrl);
  }

  getUser(id: number): Observable<User> {
    return this.http.get<User>(`${this.apiUrl}/${id}`);
  }

  createUser(user: User): Observable<User> {
    return this.http.post<User>(this.apiUrl, user);
  }

  updateUser(id: number, user: User): Observable<User> {
    return this.http.put<User>(`${this.apiUrl}/${id}`, user);
  }

  deleteUser(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  registerToCourse(studentId: number, courseDetailsId: number): Observable<any> {
    return this.http.post(`${this.apiUrl}/${studentId}/courseDetails/${courseDetailsId}`, null, { responseType: 'text' });
  }

  getRegisteredCourseDetails(studentId: number, courseId: number): Observable<CourseDetails[]> {
    return this.http.get<CourseDetails[]>(`${this.apiUrl}/${studentId}/registered-courseDetails/${courseId}`);
  }
}
